File: python/assignment.py
from abc import ABC, abstractmethod


class BankAccount:
    def __init__(self, account_number, account_holder_name, initial_balance=0.0):
        self.__account_number = account_number
        self.__account_holder_name = account_holder_name
        self.__balance = initial_balance

    def deposit(self, amount):
        if amount <= 0:
            print("Error: Deposit amount must be positive.")
            return
        self.__balance += amount
        print(
            f"Deposited GHS {amount:.2f}. New balance: GHS {self.__balance:.2f}")

    def withdraw(self, amount):
        if amount <= 0:
            print("Error: Withdrawal amount must be positive.")
            return
        if amount > self.__balance:
            print("Error: Insufficient funds.")
            return
        self.__balance -= amount
        print(
            f"Withdrew GHS {amount:.2f}. New balance: GHS {self.__balance:.2f}")

    def get_balance(self):
        return self.__balance

    def get_account_number(self):
        return self.__account_number


class Student:
    def __init__(self, student_id, name, score):
        self.__student_id = student_id
        self.__name = name
        self.score = score

    @property
    def score(self):
        return self.__score

    @score.setter
    def score(self, value):
        if not (0 <= value <= 100):
            raise ValueError(
                f"Invalid score {value}. Score must be between 0 and 100.")
        self.__score = value

    @property
    def name(self):
        return self.__name

    @property
    def student_id(self):
        return self.__student_id

    def get_grade(self):
        if self.__score >= 70:
            return "A"
        elif self.__score >= 60:
            return "B"
        elif self.__score >= 50:
            return "C"
        elif self.__score >= 45:
            return "D"
        else:
            return "F"


class Employee:
    def __init__(self, name, salary):
        self.name = name
        self.salary = salary

    def calculate_annual_salary(self):
        return self.salary * 12


class FullTimeEmployee(Employee):
    def __init__(self, name, salary, bonus):
        super().__init__(name, salary)
        self.bonus = bonus

    def calculate_annual_salary(self):
        return super().calculate_annual_salary() + self.bonus


class PartTimeEmployee(Employee):
    def __init__(self, name, hours_worked, hourly_rate):
        super().__init__(name, salary=0)
        self.hours_worked = hours_worked
        self.hourly_rate = hourly_rate

    def calculate_annual_salary(self):
        return self.hours_worked * self.hourly_rate * 52


class Vehicle:
    def __init__(self, brand):
        self.brand = brand

    def start_engine(self):
        print(f"{self.brand} engine starting...")


class Car(Vehicle):
    def start_engine(self):
        super().start_engine()
        print(f"{self.brand} car engine running smoothly.")


class Motorcycle(Vehicle):
    def start_engine(self):
        super().start_engine()
        print(f"{self.brand} motorcycle engine roaring.")


class Payment(ABC):
    @abstractmethod
    def process_payment(self, amount):
        pass


class CreditCardPayment(Payment):
    def process_payment(self, amount):
        print(f"Processing credit card payment of GHS {amount:.2f}")


class MobileMoneyPayment(Payment):
    def process_payment(self, amount):
        print(f"Processing mobile money payment of GHS {amount:.2f}")


class BankTransferPayment(Payment):
    def process_payment(self, amount):
        print(f"Processing bank transfer of GHS {amount:.2f}")


class ATMOperations(ABC):
    @abstractmethod
    def withdraw(self, amount):
        pass

    @abstractmethod
    def deposit(self, amount):
        pass

    @abstractmethod
    def check_balance(self):
        pass


class BankATM(ATMOperations):
    def __init__(self, initial_balance=0.0):
        self.__balance = initial_balance

    def deposit(self, amount):
        if amount <= 0:
            print("Invalid deposit amount.")
            return
        self.__balance += amount
        print(f"ATM: Deposited GHS {amount:.2f}.")

    def withdraw(self, amount):
        if amount <= 0:
            print("Invalid withdrawal amount.")
            return
        if amount > self.__balance:
            print("ATM: Insufficient funds.")
            return
        self.__balance -= amount
        print(f"ATM: Dispensing GHS {amount:.2f}.")

    def check_balance(self):
        print(f"ATM: Current balance is GHS {self.__balance:.2f}")


if __name__ == "__main__":

    acc = BankAccount("234908892", "Kwame Addo Mensah", 500.0)
    acc.deposit(200)
    acc.deposit(-50)
    acc.withdraw(100)
    acc.withdraw(1000)
    print(
        f"Account: {acc.get_account_number()} | Balance: GHS {acc.get_balance():.2f}")

    s = Student("22303446", "Ferdinard Afful Bentum", 75)
    print(f"{s.name} scored {s.score} Grade: {s.get_grade()}")
    s.score = 80
    print(f"Updated score: {s.score} Grade: {s.get_grade()}")
    try:
        s.score = 110
    except ValueError as e:
        print(f"Caught error: {e}")

    ft = FullTimeEmployee("Kofi Asante", 3000, 5000)
    pt = PartTimeEmployee("Abena Boateng", 20, 25)
    print(f"{ft.name} annual salary: GHS {ft.calculate_annual_salary():.2f}")
    print(f"{pt.name} annual salary: GHS {pt.calculate_annual_salary():.2f}")

    car = Car("Toyota")
    moto = Motorcycle("Yamaha")
    car.start_engine()
    moto.start_engine()

    payments = [CreditCardPayment(), MobileMoneyPayment(),
                BankTransferPayment()]
    amounts = [150.00, 75.50, 300.00]
    for payment, amount in zip(payments, amounts):
        payment.process_payment(amount)

    atm = BankATM(1000.0)
    atm.check_balance()
    atm.deposit(500)
    atm.withdraw(200)
    atm.withdraw(2000)
    atm.check_balance()

File: php/assignment.php
<?php

class BankAccount
{
    private $accountNumber;
    private $accountHolderName;
    private $balance;

    public function __construct($accountNumber, $name, $initialBalance)
    {
        $this->accountNumber = $accountNumber;
        $this->accountHolderName = $name;
        $this->balance = $initialBalance;
    }

    public function deposit($amount)
    {
        if ($amount <= 0) {
            echo "Error: Deposit amount must be positive" . "\n";
        } else {
            $this->balance += $amount;
            echo "Successfully deposited {$amount}" . "\n";
        }
    }

    public function withdraw($amount)
    {
        if ($amount > $this->balance) {
            echo "Error: Insufficient funds for withdrawal" . "\n";
        } else {
            $this->balance -= $amount;
            echo "Successfully withdrew {$amount}" . "\n";
        }
    }

    public function getBalance()
    {
        return $this->balance;
    }
    public function getAccountNumber()
    {
        return $this->accountNumber;
    }
}


class Student
{
    private $studentId;
    private $name;
    private $score;

    public function setScore($score)
    {
        if ($score >= 0 && $score <= 100) {
            $this->score = $score;
        } else {
            echo "Error: Invalid score ({$score}), Must be between 0 and 100" . "\n";
        }
    }

    public function getGrade()
    {
        if ($this->score >= 70) return "A";
        if ($this->score >= 60) return "B";
        if ($this->score >= 50) return "C";
        if ($this->score >= 45) return "D";
        return "F";
    }
}


class Employee
{
    protected $name;
    protected $salary;

    public function __construct($name, $salary)
    {
        $this->name = $name;
        $this->salary = $salary;
    }

    public function calculateAnnualSalary()
    {
        return $this->salary * 12;
    }
}

class FullTimeEmployee extends Employee
{
    private $bonus;

    public function __construct($name, $salary, $bonus)
    {
        parent::__construct($name, $salary);
        $this->bonus = $bonus;
    }

    public function calculateAnnualSalary()
    {
        return ($this->salary * 12) + $this->bonus;
    }
}

class PartTimeEmployee extends Employee
{
    private $hoursWorked;
    private $hourlyRate;

    public function __construct($name, $hours, $rate)
    {
        parent::__construct($name, 0);
        $this->hoursWorked = $hours;
        $this->hourlyRate = $rate;
    }

    public function calculateAnnualSalary()
    {
        return ($this->hoursWorked * $this->hourlyRate) * 12;
    }
}


class Vehicle
{
    protected $brand;

    public function __construct($brand)
    {
        $this->brand = $brand;
    }

    public function startEngine()
    {
        return "Starting the {$this->brand} engine: ";
    }
}

class Car extends Vehicle
{
    public function startEngine()
    {
        return parent::startEngine() . "Vroom vroom!" . "\n";
    }
}

class Motorcycle extends Vehicle
{
    public function startEngine()
    {
        return parent::startEngine() . "Brap brap!" . "\n";
    }
}


abstract class Payment
{
    abstract public function processPayment($amount);
}

class CreditCardPayment extends Payment
{
    public function processPayment($amount)
    {
        echo "Paid {$amount} via Credit Card" . "\n";
    }
}

class MobileMoneyPayment extends Payment
{
    public function processPayment($amount)
    {
        echo "Paid {$amount} via MoMo" . "\n";
    }
}


interface ATMOperations
{
    public function withdraw($amount);
    public function deposit($amount);
    public function checkBalance();
}

class BankATM implements ATMOperations
{
    private $currentBalance = 1000.0;
    public function withdraw($amount)
    {
        $this->currentBalance -= $amount;
        echo "ATM: Withdrew {$amount}" . "\n";
    }
    public function deposit($amount)
    {
        $this->currentBalance += $amount;
        echo "ATM: Deposited {$amount}" . "\n";
    }
    public function checkBalance()
    {
        echo "ATM Balance: {$this->currentBalance}" . "\n";
    }
}


$acc = new BankAccount("240454799", "Ferdinard Afful Bentum", 500);
$acc->deposit(200);
$acc->deposit(-50);
$acc->withdraw(100);
$acc->withdraw(2000);


$stu = new Student();
$stu->setScore(150);
$stu->setScore(85);
echo "Grade: " . $stu->getGrade() . "\n";


$ft = new FullTimeEmployee("Hafeez", 5000, 2000);
$pt = new PartTimeEmployee("Adam", 40, 50);
echo "FullTime Annual: {$ft->calculateAnnualSalary()}" . "\n";
echo "PartTime Annual: {$pt->calculateAnnualSalary()}" . "\n";

$myCar = new Car("Toyota");
echo $myCar->startEngine();


$payments = [new CreditCardPayment(), new MobileMoneyPayment()];
foreach ($payments as $p) {
    $p->processPayment(100);
}


$atm = new BankATM();
$atm->checkBalance();
$atm->deposit(500);
